drivers/eeprom/24xx32a.c

```diff
@@ -0,0 +1,221 @@
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdbool.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "24xx32a.h"
+#include "no_os_eeprom.h"
+#include "no_os_i2c.h"
+#include "no_os_error.h"
+#include "no_os_delay.h"
+
+/******************************************************************************/
+/************************** Functions Definitions *****************************/
+/******************************************************************************/
+
+/**
+ * @brief 	Initialize the 24XX32A EEPROM
+ * @param	desc - EEPROM descriptor
+ * @param	param - EEPROM init parameters
+ * @return	0 in case of success, negative error code otherwise
+ */
+int32_t eeprom_24xx32a_init(struct no_os_eeprom_desc **desc,
+			    const struct no_os_eeprom_init_param *param)
+{
+	int32_t ret;
+	struct no_os_eeprom_desc *eeprom_desc;
+	struct eeprom_24xx32a_dev *eeprom_dev;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	eeprom_desc = (struct no_os_eeprom_desc*)calloc(1, sizeof(*eeprom_desc));
+	if (!eeprom_desc)
+		return -ENOMEM;
+
+	eeprom_desc->device_id = param->device_id;
+
+	eeprom_dev = (struct eeprom_24xx32a_dev*)calloc(1, sizeof(*eeprom_dev));
+	if (!eeprom_dev) {
+		ret = -ENOMEM;
+		goto error_eeprom_dev;
+	}
+
+	ret = no_os_i2c_init((struct no_os_i2c_desc **)&eeprom_dev->i2c_desc,
+			     ((struct eeprom_24xx32a_init_param *)param->extra)->i2c_init);
```

Comment:
The first cast is not needed. The second may be replaced by declaring a variable of type `struct eeprom_24xx32a_init_param *` and assigning the `param->extra` to that. It's generally preferred to do it the way I described, since the code will be less convoluted.

Reply:
Make sense, will update code accordingly.